src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs

```diff
@@ -12,13 +16,81 @@ internal static class WebInputSources
         /// </summary>
         public static ImmutableHashSet<SourceInfo> SourceInfos { get; }
 
+        /// <summary>
+        /// Cached information if the specified symbol is a Asp.Net Core Controller: (compilation) -> ((class symbol) -> (is Controller))
+        /// </summary>
+        private static readonly BoundedCacheWithFactory<Compilation, ConcurrentDictionary<ISymbol, bool>> s_classIsControllerByCompilation =
+            new BoundedCacheWithFactory<Compilation, ConcurrentDictionary<ISymbol, bool>>();
+
+        /// <summary>
+        /// Cached information if the specified symbol is a Asp.Net Core Action: (compilation) -> ((method symbol) -> (is Action))
+        /// </summary>
+        private static readonly BoundedCacheWithFactory<Compilation, ConcurrentDictionary<ISymbol, bool>> s_methodIsActionByCompilation =
+            new BoundedCacheWithFactory<Compilation, ConcurrentDictionary<ISymbol, bool>>();
+
         /// <summary>
         /// Statically constructs.
         /// </summary>
         static WebInputSources()
         {
             var sourceInfosBuilder = PooledHashSet<SourceInfo>.GetInstance();
 
+            sourceInfosBuilder.AddSourceInfo(
+                // checking all System.Object derived types is expensive, so it first checks if MicrosoftAspNetCoreMvcControllerBase is resolvable
+                WellKnownTypeNames.MicrosoftAspNetCoreMvcControllerBase,
+                WellKnownTypeNames.SystemObject,
+                 new ParameterMatcher[]{
+                    (parameter, wellKnownTypeProvider) => {
+                        // it can be expensive to do all the machinery on every parameter reference access
+                        var methodCache = s_methodIsActionByCompilation.GetOrCreateValue(wellKnownTypeProvider.Compilation, (compilation) => new ConcurrentDictionary<ISymbol, bool>());
+                        if (methodCache.TryGetValue(parameter.ContainingSymbol, out bool isAction))
+                            return isAction;
+
+                        if (!(parameter.ContainingSymbol is IMethodSymbol methodSymbol))
+                        {
+                            methodCache.TryAdd(parameter.ContainingSymbol, false);
+                            return false;
+                        }
+
+                        var classCache = s_classIsControllerByCompilation.GetOrCreateValue(wellKnownTypeProvider.Compilation, (compilation) => new ConcurrentDictionary<ISymbol, bool>());
+                        if (!classCache.TryGetValue(methodSymbol.ContainingSymbol, out bool isController))
```

Comment:
I think the `out bool isController` value needs to be checked. Should probably add a unit test that has a non-controller with multiple methods referencing parameters.

Reply:
Good catch!